src/Mod/Part/App/Geometry.cpp

```diff
@@ -1596,19 +1596,9 @@ void GeomBSplineCurve::Trim(double u, double v)
     };
 
     try {
-        if(!isPeriodic()) {
-            splitUnwrappedBSpline(u, v);
-        }
-        else { // periodic
-            if( v < u ) { // wraps over origin
-                v = v + 1.0; // v needs one extra lap (1.0)
-
-                splitUnwrappedBSpline(u, v);
-            }
-            else {
-                splitUnwrappedBSpline(u, v);
-            }
-        }
+        if (isPeriodic() && (v - u  <= Precision::PConfusion()))
+            v = v + myCurve->LastParameter() - myCurve->FirstParameter(); // v needs one extra lap
+        splitUnwrappedBSpline(u, v);

```

Comment:
I am not sure I understand correctly the motivation for the "<= Precision::PConfusion()".

A BSpline is wrapped if the v parameter is less than the u parameter. A BSpline where v is a positive amount (Precision::PConfusion()) greater than u is not wrapped. In such a case v is greater than u. Am I missing anything?

Then, to unwrap it, without changing it, v needs a full lap (1.0). However, it appears to me that LastParameter()-FirstParameter() will only be one lap if the Bpsline has not been split before (if it is a full "turn"). I have not checked, maybe the split of a periodic b-spline is no longer a periodic b-spline (so the code is not executed). Even if that is the case, why not use 1.0 addition instead, which clearly conveys the intention? Maybe I am missing something?

Reply:
> I am not sure I understand correctly the motivation for the "<= Precision::PConfusion()".
> 
> A BSpline is wrapped if the v parameter is less than the u parameter. A BSpline where v is a positive amount (Precision::PConfusion()) greater than u is not wrapped. In such a case v is greater than u. Am I missing anything?

`PConfusion` refers to a small epsilon because of roundoff errors. Let's consider a periodic spline with period 1.0. If I want to split it at point 0.5, and I've for some reason provided two different values (maybe because these values had to be computed by different clicks, say---which is not our case in split, but this change is in part and general). Then, maybe u=0.5 and v=0.499999997, or v=0.50000000001. In the first case, with just "v<=u" the wrapping will be considered. However, in the latter case, we will get just a small arc from 0.5 to 0.50000000001.

Note that if I had needed to check for strictly "v<u", I would have instead used `(v - u  <= -Precision::PConfusion())`. Also, here I expect that `<=` can be replaced by `<`.

> Then, to unwrap it, without changing it, v needs a full lap (1.0). However, it appears to me that LastParameter()-FirstParameter() will only be one lap if the Bpsline has not been split before (if it is a full "turn"). 
> ...why not use 1.0 addition instead, which clearly conveys the intention? Maybe I am missing something?

The `LastParameter()-FirstParameter()` is for the most general case. I don't expect it would happen organically, but I could, for example, split a nonperiodic spline at 0.6, delete the second part, then join the ends of the first part. We could enforce that every periodic spline in FC should always have a period of 1.0, but I don't see any particular good in that.

> I have not checked, maybe the split of a periodic b-spline is no longer a periodic b-spline (so the code is not executed). 

The code is executed as if for a periodic b-spline. Only after the OCC algorithm is executed does it become non-periodic.

Reply:
> The `LastParameter()-FirstParameter()` is for the most general case. I don't expect it would happen organically, but I could, for example, split a nonperiodic spline at 0.6, delete the second part, then join the ends of the first part. We could enforce that every periodic spline in FC should always have a period of 1.0, but I don't see any particular good in that.

I thought OCC would always enforce 1.0 for periodic B-Splines (including the non-periodic to periodic case). If the periodic B-spline changes after OCC upon trimming to non-periodic, then it is acceptable to me to have the subtraction of parameters.

> `PConfusion` refers to a small epsilon because of roundoff errors.

Agreed.

> Let's consider a periodic spline with period 1.0. If I want to split it at point 0.5, and I've for some reason provided two different values (maybe because these values had to be computed by different clicks, say---which is not our case in split, but this change is in part and general). Then, maybe u=0.5 and v=0.499999997, or v=0.50000000001. In the first case, with just "v<=u" the wrapping will be considered. However, in the latter case, we will get just a small arc from 0.5 to 0.50000000001.

But, isn't this a case of wrong input to the Part function? 

In a general case, if I am telling this function I want to trim between u=0.5 and v=0.5000000000001, I expect, as a user of this function, that it produces very small arc (for it is what I am asking for). Getting a long arc is not what I am asking for. That would look to me as unexpected behaviour (when looking at the input data I am providing).

I can follow the argument that in your case, rounding playing against your will, you may be feeding the wrong numbers to the function (the ones of the extra short arc). However, the function should not try to guess whether the user may have wanted something else that what he is asking for. In other words, if you expect the large arc, the client code calling the function has to ensure it provides the right numbers (that it corrects wrong input data before calling the function). In your example, being a mouse input, input data validation (at client code) should be responsible for this.

> Note that if I had needed to check for strictly "v<u", I would have instead used `(v - u <= -Precision::PConfusion())`

That would have the contrary effect, that would also be undefined behaviour.

> Also, here I expect that `<=` can be replaced by `<`

When the trimming distance (in term of parameter bounds) is under Confusion, the OCC algorithm will produce an exception. This will be caught and trigger a CADKernelError for the client code to handle. If the client code wants to avoid receiving an exception, it needs to ensure it is providing values that do not produce one. 

I tend to think that if a user is asking for an impossible getting an exception/error is what he should expect. It would be unexpected to get a valid result, when an invalid result is expected.

Reply:
> > Let's consider a periodic spline with period 1.0. If I want to split it at point 0.5, and I've for some reason provided two different values (maybe because these values had to be computed by different clicks, say---which is not our case in split, but this change is in part and general). Then, maybe u=0.5 and v=0.499999997, or v=0.50000000001. In the first case, with just "v<=u" the wrapping will be considered. However, in the latter case, we will get just a small arc from 0.5 to 0.50000000001.
> 
> But, isn't this a case of wrong input to the Part function?
> 
> In a general case, if I am telling this function I want to trim between u=0.5 and v=0.5000000000001, I expect, as a user of this function, that it produces very small arc (for it is what I am asking for). Getting a long arc is not what I am asking for. That would look to me as unexpected behaviour (when looking at the input data I am providing).

Off the top of my head, I can think of one way this can happen. Say we implement an "arc" tool that takes an arc on the curve between two points. A user quickly clicks around the same point to get the same effect as "splitting" the loop. Because of the error this can go any way. 

A slightly more practical case maybe trying to find union/intersection of tangential faces, though in that case we probably will also need a way to replace segments by just points. I'll skip this example at least for now.
 
> > Note that if I had needed to check for strictly "v<u", I would have instead used `(v - u <= -Precision::PConfusion())`
> 
> That would have the contrary effect, that would also be undefined behaviour.

Note that u and/or v can be negative. We just want to ensure that "v<u" or "v-u<0" _without any doubt_. So we sacrifice that much amount of confusion to ensure that they are definitely not too close to each other. 

> I tend to think that if a user is asking for an impossible getting an exception/error is what he should expect. It would be unexpected to get a valid result, when an invalid result is expected.

I'm fairly certain that the things requested here are valid.

Reply:
> > We just want to ensure that "v<u" or "v-u<0" without any doubt.
> 
> If I do "if( v<u)" and it returns true, I have no doubt at all that v is smaller than u.

You can ensure that the _provided_ v is smaller than the _provided_ u, but the _intended_ values maybe too close to each other.

> 
> We must be having a communication problem here. I will try again with different wording.

It may be difference in opinion as to where the check for confusion should be done. The check for "v < u" is already done here before I changed the file, so I don't expect things to break anywhere else.

>     1. There are many user input cases susceptible of producing values of v and u that the user does not want. This I can agree. The portion of the application getting this input from the user is responsible for validating user input data. It can do as many checks as necessary and condition this data. It must decide what it wants to trim to get the effect it wants to achieve. Then it can use those in this trim function.

I'm not sure if it's a good idea to depend on all "callers" to do this. As such then we can even impose that "u<v" and assume that is true. 

>     2. I cannot understand why a function that offers the functionality to trim a bspline between u and v should "sacrifice" part of its working range and wrap something that does not need to be wrapped. I cannot understand why if I am asking to trim between 0.5-Confusion and 0.5, the function should return a trim between 0.5-Conf and 1.5.

This is unfortunately an eternal problem in computational geometry. Thing is, if we do ask a trim between 0.5-Confusion and 0.5, OCC is not guaranteed to provide the intended result. Pick any of the solved bugs in FC tagged "OCC", and take a look at the commit that fixed it. I am certain that most of the fixes involved incorporating confusion somewhere in their checks.

If someone really needs a higher precision than `Confusion`, this will most likely involve extensive changes in the algorithm and/or performance penalty because of the higher precision checks.

As such, since we are directly passing these parameters to OCC, the tolerance has to be at least as much OCC allows. Using a more sophisticated solver, we may eventually get to numbers like `0.000001*Precision::Confusion` but even then this check will be needed. The price for not making this sacrifice is instability.

Also take a look at https://dev.opencascade.org/doc/occt-7.6.0/refman/html/class_geom_convert.html#a6e7ddc155168bc9b9b2f21d4d5295d42. It clearly stats "Raised if Abs (FromU1 - ToU2) <= ParametricTolerance", i.e. if they are too close.

> BTW, the wrapped parameters are confusion in excess of one "lap"...

Not sure how the behavior would change in this case, but I expect this should be fine so long as the difference past the "lap" is not too much. Again, from the reference "Raised if FromU1 or ToU2 are out of the parametric bounds of the curve (_The tolerance criterion is ParametricTolerance_)", so I expect that there are safeguards against it.